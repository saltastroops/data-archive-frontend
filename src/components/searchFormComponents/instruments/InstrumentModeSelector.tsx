import * as React from "react";
import MultiSelectField, {
  AllOption
} from "../../basicComponents/MultiSelectField";

// MODES
export const COMMON_RSS_BVIT_MODES = ["Imaging"];
export const RSS_MODES = [
  "Fabry Perot",
  // "FP Polarimetry",
  "MOS",
  // "MOS Polarimetry",
  "Polarimetric Imaging",
  "Spectropolarimetry",
  "Spectroscopy"
];
export const BVIT_MODES = ["Streaming"];

interface IInstrumentModeSelectorProps {
  instrumentModes: string[];
  selectedInstruments?: string[];
  selectedTelescopes?: string[];
  onSelect: (value: any) => void;
}

const modesToDisplay = (instruments: string[], telescopes: string[]) => {
  let modes: string[] = [];
  if (
    instruments.length === 0 ||
    (instruments.some(inst => inst === "All") &&
      telescopes.some(t => t === "All" || t === "SALT"))
  ) {
    return RSS_MODES.concat(BVIT_MODES).concat(COMMON_RSS_BVIT_MODES);
  }
  if (telescopes.some(t => t === "1.9 m" || t === "Lesedi")) {
    return [];
  }
  if (instruments.some(inst => inst === "RSS")) {
    modes = modes.concat(RSS_MODES);
  }
  if (instruments.some(inst => inst === "BVIT")) {
    modes = modes.concat(BVIT_MODES);
  }
  if (instruments.some(inst => inst === "BVIT" || inst === "RSS")) {
    modes = modes.concat(COMMON_RSS_BVIT_MODES);
  }

  return modes;
};
/**
 * A form for selecting ModeSelector-related search parameters.
 */
const InstrumentModeSelector = (props: IInstrumentModeSelectorProps) => {
  const {
    onSelect,
    selectedInstruments,
    selectedTelescopes,
    instrumentModes
  } = props;

  // Function for handling change events
  const select = (e: React.SyntheticEvent<HTMLSelectElement>) => {
    const values = e.currentTarget.selectedOptions;
    onSelect({
      instrumentModes: Array.from(values).map((t: any) => {
        return t.text;
      })
    });
  };
  const MODE_TO_DISPLAY = modesToDisplay(
    selectedInstruments || [],
    selectedTelescopes || []
  );

  return (
    <div>
      <p>Mode</p>
      <MultiSelectField
        name={"instrumentModes"}
        onChange={select}
        value={instrumentModes || ["All"]}
      >
        <AllOption />
        {MODE_TO_DISPLAY.sort().map(i => (
          <option key={i} value={i}>
            {i}
          </option>
        ))}
      </MultiSelectField>
    </div>
  );
};

export default InstrumentModeSelector;
