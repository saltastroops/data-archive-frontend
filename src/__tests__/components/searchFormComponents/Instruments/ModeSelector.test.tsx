import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import * as React from "react";
import InstrumentModeSelector from "../../../../components/searchFormComponents/instruments/InstrumentModeSelector";

const onSelect = jest.fn() as any;
describe("Mode Selector", () => {
  it("should be defined", () => {
    expect(
      mount(<InstrumentModeSelector instrumentModes={[]} onSelect={onSelect} />)
    ).toBeDefined();
  });

  it("should render correctly", () => {
    // Use mount instead of shallow for better snapshots
    expect(
      toJson(
        mount(
          <InstrumentModeSelector instrumentModes={[]} onSelect={onSelect} />
        )
      )
    ).toMatchSnapshot();
  });
});
