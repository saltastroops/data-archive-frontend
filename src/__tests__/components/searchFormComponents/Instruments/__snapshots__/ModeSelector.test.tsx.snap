// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mode Selector should render correctly 1`] = `
.c0 {
  width: 95%;
}

<ModeSelector
  modes={Array []}
  onSelect={[MockFunction]}
>
  <div>
    <p>
      Mode
    </p>
    <MultiSelectField
      name="modes"
      onChange={[Function]}
      value={Array []}
    >
      <styled.div
        className="control select is-multiple "
      >
        <StyledComponent
          className="control select is-multiple "
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-bdVaJa",
                "isStatic": true,
                "lastClassName": "c0",
                "rules": Array [
                  "
  width: 95%;
",
                ],
              },
              "displayName": "styled.div",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-bdVaJa",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <div
            className="control select is-multiple  c0"
          >
            <select
              multiple={true}
              name="modes"
              onChange={[Function]}
              size={6}
              style={
                Object {
                  "width": "216px",
                }
              }
              value={Array []}
            >
              <AllOption>
                <option
                  value="All"
                >
                  All
                </option>
              </AllOption>
              <option
                key="Fabry Perot"
                value="Fabry Perot"
              >
                Fabry Perot
              </option>
              <option
                key="Imaging"
                value="Imaging"
              >
                Imaging
              </option>
              <option
                key="MOS"
                value="MOS"
              >
                MOS
              </option>
              <option
                key="Polarimetric imaging"
                value="Polarimetric imaging"
              >
                Polarimetric imaging
              </option>
              <option
                key="Spectropolarimetry"
                value="Spectropolarimetry"
              >
                Spectropolarimetry
              </option>
              <option
                key="Spectroscopy"
                value="Spectroscopy"
              >
                Spectroscopy
              </option>
              <option
                key="Streaming"
                value="Streaming"
              >
                Streaming
              </option>
            </select>
          </div>
        </StyledComponent>
      </styled.div>
    </MultiSelectField>
  </div>
</ModeSelector>
`;
